wdl-analysis/src/queue.rs

```diff
@@ -21,146 +21,238 @@ use tokio::sync::oneshot;
+pub struct AnalysisQueue<Progress, Context, Return, Validator> {
     /// The document graph maintained by the analysis queue.
     graph: Arc<RwLock<DocumentGraph>>,
     /// The handle to the tokio runtime for blocking on async tasks.
     tokio: Handle,
     /// The HTTP client to use for fetching documents.
     client: Client,
     /// The progress callback to use.
-    progress: Arc<P>,
-    /// A marker for the `R` type.
-    marker: PhantomData<R>,
+    progress: Arc<Progress>,
     /// The validator callback to use.
-    validator: Arc<V>,
-    /// The changes callback to use.
-    changes: C,
+    validator: Arc<Validator>,
+    /// A marker for the `Context` and `Return` types.
+    marker: PhantomData<(Context, Return)>,
 }
 
-impl<P, R, V, C> AnalysisQueue<P, R, V, C>
+impl<Progress, Context, Return, Validator> AnalysisQueue<Progress, Context, Return, Validator>
 where
-    P: Fn(ProgressKind, usize, usize, Option<String>) -> R + Send + Sync + 'static,
-    R: Future<Output = ()>,
-    V: Fn() -> Validator + Send + Sync + 'static,
-    C: Fn(&Url) -> Option<DocumentChange>,
+    Progress: Fn(Context, ProgressKind, usize, usize) -> Return + Send + 'static,
+    Context: Send + Clone,
+    Return: Future<Output = ()>,
+    Validator: Fn() -> wdl_ast::Validator + Send + Sync + 'static,
 {
     /// Constructs a new analysis queue.
-    pub fn new(tokio: Handle, progress: P, validator: V, changes: C) -> Self {
+    pub fn new(tokio: Handle, progress: Progress, validator: Validator) -> Self {
         Self {
             graph: Default::default(),
             tokio,
             progress: Arc::new(progress),
             marker: PhantomData,
             client: Default::default(),
             validator: Arc::new(validator),
-            changes,
         }
     }
 
     /// Runs the analysis queue.
-    pub fn run(&self, mut receiver: UnboundedReceiver<Request>) {
+    pub fn run(&self, mut receiver: UnboundedReceiver<Request<Context>>) {
         log::info!("analysis queue has started");
 
         while let Some(request) = self.tokio.block_on(receiver.recv()) {
             match request {
-                Request::Analyze(AnalyzeRequest {
+                Request::Add(AddRequest {
                     documents,
-                    context,
                     completed,
                 }) => {
                     let start = Instant::now();
                     log::info!(
-                        "received request to analyze {count} document(s)",
+                        "received request to add {count} document(s) to the graph",
                         count = documents.len()
                     );
 
-                    self.analyze(documents, context, completed);
+                    self.add_documents(documents);
 
                     log::info!(
-                        "analysis request completed in {elapsed:?}",
+                        "request to add documents completed in {elapsed:?}",
```

Comment:
Considering that we're getting to a level of maturity where we want this kind of profiling, I wonder if it makes sense to switch to leveraging `tracing` in these instances. Shouldn't hold up the PR, but we may thank ourselves later rather than having to go back and add it in.

Reply:
I think that would be a great follow-up PR, definitely put in sooner than later.